src/main/java/io/pravega/connectors/flink/FlinkPravegaInputFormat.java

```diff
@@ -0,0 +1,179 @@
+        Preconditions.checkNotNull(deserializationSchema, "deserializationSchema");
+
+        this.controllerURI = controllerURI;
+        this.scopeName = scope;
+        this.deserializationSchema = deserializationSchema;
+        this.streamNames = streamNames;
+        this.startTime = startTime;
+        this.readerGroupName = generateRandomReaderGroupName();
+
+        // TODO: This will require the client to have access to the pravega controller and handle any temporary errors.
+        ReaderGroupManager.withScope(scopeName, controllerURI)
+                .createReaderGroup(this.readerGroupName, ReaderGroupConfig.builder().startingTime(startTime).build(),
+                        streamNames);
+    }
+
+    // ------------------------------------------------------------------------
+    //  User configurations
+    // ------------------------------------------------------------------------
+
+    /**
+     * Gets the timeout for the call to read events from Pravega. After the timeout
+     * expires (without an event being returned), another call will be made.
+     *
+     * <p>This timeout is passed to {@link EventStreamReader#readNextEvent(long)}.
+     *
+     * @param eventReadTimeout The timeout, in milliseconds
+     */
+    public void setEventReadTimeout(long eventReadTimeout) {
+        Preconditions.checkArgument(eventReadTimeout > 0, "timeout must be >= 0");
+        this.eventReadTimeout = eventReadTimeout;
+    }
+
+    /**
+     * Gets the timeout for the call to read events from Pravega.
+     *
+     * <p>This timeout is the value passed to {@link EventStreamReader#readNextEvent(long)}.
+     *
+     * @return The timeout, in milliseconds
+     */
+    public long getEventReadTimeout() {
+        return eventReadTimeout;
+    }
+
+    // ------------------------------------------------------------------------
+    //  Input split life cycle methods
+    // ------------------------------------------------------------------------
+
+    @Override
+    public void open(GenericInputSplit split) throws IOException {
+        super.open(split);
+
+        // build a new reader for each input split
+        this.pravegaReader = createPravegaReader(
+                this.scopeName,
+                this.controllerURI,
+                getRuntimeContext().getTaskNameWithSubtasks(),
+                this.readerGroupName,
+                this.deserializationSchema,
+                ReaderConfig.builder().build());
+    }
+
+    @Override
+    public boolean reachedEnd() throws IOException {
+        // look ahead to see if we have reached the end of input
+        try {
+            this.lastReadAheadEvent = pravegaReader.readNextEvent(eventReadTimeout);
+        } catch (Exception e) {
+            throw new IOException("Failed to read next event.", e);
+        }
+
+        // TODO this "end of input" marker is too brittle, as the timeout could easily be a temporary hiccup;
```

Comment:
See pravega/pravega#1916. We can fix this here once we complete the work of that issue.